Automattic/wp-calypso
# QueryPostTypes: clean up the component and make it faster

This patch does two bloat-reducing and performance-improving fixes to `QueryPostTypes`:

First commit improves how the component detects change in the data it depends on and decides that it needs to issue a new request. The check now almost always happens inside `react-redux` `connect` call and very rarely triggers a React state update.

Second commit migrates the actual execution of the request to data layer. Now we no longer need to maintain the `requesting` flag in Redux state. That allows us to remove several action types, reducers and selectors. And also decreases the number of action dispatches on initial load and/or site switch.

Read the commit messages for more details.

This is a fairly typical example of how Calypso wastes many CPU cycles executing inefficient code. With a large Redux tree and many connected components, action dispatches are expensive: many tiny reducers need to be executed, and also many `mapStateToProps` calls.

**How to test:**
1. Load Calypso and/or switch to another site.
2. Verify that a `GET /rest/v1.1/sites/${siteId}/post-types` was sent to the server and that the response was received into `state.postTypes.items[siteId]`.
3. While in "My Sites", the "Manage" section of the sidebar should show items for specific post types if supported by the site, e.g., "Testimonials".
